drivers/clock_control/clock_control_nrf2_fll16m.c

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_fll16m
+
+#include "clock_control_nrf2_common.h"
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/clock_control/nrf_clock_control.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(clock_control_nrf2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+#include <hal/nrf_lrcconf.h>
+
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	     "multiple instances not supported");
+
+#define FLAG_HFXO_STARTED BIT(FLAGS_COMMON_BITS)
+
+#define FLL16M_MODE_OPEN_LOOP   0
+#define FLL16M_MODE_CLOSED_LOOP 1
+#define FLL16M_MODE_BYPASS      2
+
+#define HFXO_NODE DT_NODELABEL(hfxo)
+#define LFXO_NODE DT_NODELABEL(lfxo)
+#define HFXO_ACCURACY \
+	NRF_CLOCK_CONTROL_ACCURACY_PPM(DT_PROP(HFXO_NODE, accuracy_ppm))
+
+static const struct clock_options {
+	uint16_t accuracy;
+	uint8_t mode;
+} clock_options[] = {
+	{
+		.accuracy = NRF_CLOCK_CONTROL_ACCURACY_PPM(20000),
+		.mode = FLL16M_MODE_OPEN_LOOP,
+	},
+#if DT_NODE_HAS_STATUS(LFXO_NODE, okay)
+	{
+		.accuracy = NRF_CLOCK_CONTROL_ACCURACY_PPM(5000),
+		.mode = FLL16M_MODE_CLOSED_LOOP,
+	},
+#endif
+	{
+		.accuracy = HFXO_ACCURACY,
+		.mode = FLL16M_MODE_BYPASS,
+	},
+};
+static const uint16_t max_accuracy = HFXO_ACCURACY;
+
+struct fll16m_dev_data {
+	STRUCT_CLOCK_CONFIG(fll16m, ARRAY_SIZE(clock_options)) clk_cfg;
+	struct onoff_client hfxo_cli;
+};
+
+struct fll16m_dev_config {
+	uint32_t fixed_frequency;
+};
+
+static void activate_fll16m_mode(struct fll16m_dev_data *dev_data, uint8_t mode)
+{
+	/* TODO: change to nrf_lrcconf_* function when such is available. */
+	NRF_LRCCONF010->CLKCTRL[0].SRC = mode;
```

Comment:
While debugging a clock issue with @KAGA164 I realized we've forgotten to manage LRCCONF's power. In some scenarios, it results in losing clock mode requests.

Setting `mode` to any other value than the hardware default (open-loop) should be preceded by `NRF_LRCCONF010->POWERON.MAIN = 1`. Setting `mode` to the hardware default should be followed by resetting `POWERON.MAIN`.

Please note, that other components could require setting POWERON.ACT or POWER.MAIN independently. One example is HFXO clock_control instance. So accessing LRCCONF010->POWERON.MAIN requires a multi-client helper (another on-off instance?)

Reply:
Added in new PR